ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding_llama_fused_qk/device/rotary_embedding_llama_fused_qk_device_operation.cpp

```diff
@@ -0,0 +1,105 @@
+    using namespace tt::constants;
+    TT_FATAL(input_tensors.size() == 5, "Error");
+    const auto& q_input_tensor = input_tensors.at(0);
+    const auto& k_input_tensor = input_tensors.at(1);
+    const auto& cos = input_tensors.at(2);
+    const auto& sin = input_tensors.at(3);
+    const auto& trans_mat = input_tensors.at(4);
+
+    auto ref_device = q_input_tensor.device();
+    for (const auto& input : input_tensors) {
+        TT_FATAL(input.storage_type() == StorageType::DEVICE || input.storage_type() == StorageType::MULTI_DEVICE, "Operands to rotary embedding need to be on device!");
+        TT_FATAL(input.buffer() != nullptr, "Operands to rotary embedding need to be allocated in buffers on device!");
+        TT_FATAL(input.device() == ref_device, "Operands to rotary embedding need to be on same device!");
+        TT_FATAL((input.get_layout() == Layout::TILE), "Inputs to rotary embedding must be tilized");
+        TT_FATAL((input.memory_config().memory_layout == TensorMemoryLayout::HEIGHT_SHARDED), "inputs for RoPE must be HEIGHT_SHARDED.");
+        TT_FATAL((input.get_dtype() == DataType::BFLOAT16), "Inputs to rotary embedding must be bfloat16");
+    }
+
+    // Check for decode mode
+    TT_FATAL(q_input_tensor.get_logical_shape()[0] == 1 && k_input_tensor.get_logical_shape()[0] == 1, "rotary_embedding_llama_fused_qk currently only supports deocde mode qith seq_len=1.");
+
+    TT_FATAL(q_input_tensor.get_logical_shape()[-1] == k_input_tensor.get_logical_shape()[-1], "Q input tensor and K input tensor must have same head dimensions");
+    uint32_t head_dim = q_input_tensor.get_logical_shape()[-1];
+    TT_FATAL(head_dim <= 128 || std::get<ttnn::WormholeComputeKernelConfig>(this->compute_kernel_config).fp32_dest_acc_en == false, "If head_dim is > 128, fp32_dest_acc_en must be False");
+
+    // Check that head_dim is a multiple of 32
+    TT_FATAL(head_dim % TILE_WIDTH == 0, "Head dim must be a multiple of TILE_WIDTH");
+
+    TT_FATAL(q_input_tensor.memory_config().memory_layout == this->q_output_mem_config.memory_layout, "Q Input tensor and Q output tensor must have same memory layout");
+    TT_FATAL(k_input_tensor.memory_config().memory_layout == this->k_output_mem_config.memory_layout, "K Input tensor and K output tensor must have same memory layout");
+
+    // check that q and k have same batch size and lesser that equal to 32
+    uint32_t q_batch_size = q_input_tensor.get_logical_shape()[1];
+    uint32_t k_batch_size = k_input_tensor.get_logical_shape()[1];
+    TT_FATAL(q_batch_size == k_batch_size, "Q and K must have the equal batch size");
+    TT_FATAL(q_batch_size <= 32, "Q and K must have batch size less than or equal to 32, due to parallelization over core-grid of 64");
+    uint32_t q_num_cores = q_input_tensor.shard_spec()->grid.bounding_box().grid_size().x * q_input_tensor.shard_spec()->grid.bounding_box().grid_size().y;
```

Comment:
can you add a check that Q and K are non-overlapping

Reply:
Thanks for the review!
Will add that validation.